feature request: in MMM, sort credentials by modified date feature request good first issue help wanted #207

Sriep opened the pull request:
Fix for 207.
In debug mode it will eventually assert in QAbstractItemModelPrivate::removePersistentIndexData, while deleting a QPersistentModelIndex. This occurs while handling a mouse click immediately following a sort, in QItemSelectionModel::setCurrentIndex while a copy of the previous index is being deleted.

CredentialView :
Add code so view knows the current login item. 
Handle layoutchanged signal after sorting, so the current item can be reset to the previous value.
Show header and enable sorting. Activating old sort by name code and new sort by date code.

CredentialModelFilter:
Add sort by date to lessThan. 
Override sort virtual function to remember sort order for lessThan.

CredentialModel:
Change number of columns to 2.
Modify data to return update date for new column row, and check column for other roles.
Add headerData function to return header text.

ItemDelegate:
Change so does not display date for login items.

TreeItem:
ServiceItem:
Add bestUpdateDate function to return the latest or earliest update date for item and/or subitems, dependant on sort order.

azubieta commented:
@Sriep I was testing your changes, there is a segfault when:
- enter CMM
- an item is selected
- exit CMM
- enter again in CMM
this is due CredentialView::m_pCurrentServiceItem not being reset to nullptr. It could be fixed by reseting it every time a model is set.

Also the current selected index is not restored properly, you should use the saved value to search the new position of the previously selected index. Otherwise it's set to another item.

Finally, the delete login functionality and the save/discard login changes functionalities are not working.

Sriep commented:
Thanks azubieta. I'll look at these issues and push again.

Sriep commented:
Code still asserts after select row after two successive sorts.  Selecting a row after sorting prevents the assert. Still looking into this.

Removed segfault when leaving CMM.

Allow save and delete of logins. Changed so that when deleting a row, the row is deselected before being deleted. This prevents deleted information being accessed.

Store login and service item before a sort. Use to set current index after sorting.

limpkin commented:
moolticute\src\ItemDelegate.cpp:-1: In member function 'void ItemDelegate::paintFavorite(QPainter*, const QStyleOptionViewItem&, int) const':
moolticute\src\ItemDelegate.cpp:58: warning: unused variable 'optionrecheight' [-Wunused-variable]
     int optionrecheight = option.rect.height();
moolticute\src\ItemDelegate.cpp:59: warning: variable 'topleft' set but not used [-Wunused-but-set-variable]
     QPoint topleft = option.rect.topLeft();
moolticute\src\ItemDelegate.cpp:82: warning: unused parameter 'iFavorite' [-Wunused-parameter]
                               , int iFavorite) const

Sriep commented on src/CredentialModel.cpp:
```diff
@@ -35,12 +36,22 @@ QVariant CredentialModel::data(const QModelIndex &idx, int role) const
+    if (role == Qt::DisplayRole && idx.column() == 0)
+    {
+        if (pLoginItem != nullptr)
+            return  "         " + pItem->name();
```
It was so there is room for favourite star or arrow. I can either move repositioning text to delegate or move favourite star and arrow drawing to model as DecorationRole.

raoulh replied:
The text can be repositioned in the CredentialView when the icon is displayed. no need for DecorationRole

raoulh commented on src/ItemDelegate.cpp:
```diff
@@ -103,25 +103,22 @@ void ItemDelegate::paintLoginItem(QPainter *painter
+            painter->setFont(font);
+
+            QString indent(8, ' ');
+            painter->drawText(option.rect, Qt::AlignVCenter, indent + pLoginItem->name());
```
Adding spaces here instead of the model is not what I commented...
Here you need to draw the text by moving the `option.rect` and not by adding spaces...

Sriep replied:
Ok.
